executor/executor_runner.h

```diff
@@ -0,0 +1,780 @@
+			HandleCompletion(status);
+			if (out_of_requests)
+				wait_start_ = current_time_ms();
+		} else {
+			debug("got data on response pipe in wrong state %d\n", state_);
+			return false;
+		}
+		return true;
+	}
+
+	bool ReadOutput()
+	{
+		const size_t kChunk = 1024;
+		output_.resize(output_.size() + kChunk);
+		ssize_t n = read(stdout_pipe_, output_.data() + output_.size() - kChunk, kChunk);
+		if (n < 0) {
+			if (errno == EINTR || errno == EAGAIN)
+				return true;
+			fail("proc stdout read failed");
+		}
+		output_.resize(output_.size() - kChunk + n);
+		if (n == 0) {
+			debug("proc %d: output pipe EOF\n", id_);
+			return false;
+		}
+		if (flag_debug) {
+			output_.resize(output_.size() + 1);
+			debug("proc %d: got output: %s\n", id_, output_.data() + debug_output_pos_);
+			output_.resize(output_.size() - 1);
+			debug_output_pos_ = output_.size();
+		}
+		return true;
+	}
+};
+
+// Runner manages a set of test subprocesses (Proc's), receives new test requests from the manager,
+// and dispatches them to subprocesses.
+class Runner
+{
+public:
+	Runner(Connection& conn, const char* name, const char* bin)
+	    : conn_(conn),
+	      name_(name)
+	{
+		size_t num_procs = Handshake();
+		int max_signal_fd = max_signal_ ? max_signal_->FD() : -1;
+		int cover_filter_fd = cover_filter_ ? cover_filter_->FD() : -1;
+		for (size_t i = 0; i < num_procs; i++)
+			procs_.emplace_back(new Proc(conn, bin, i, max_signal_fd, cover_filter_fd,
+						     slowdown_, syscall_timeout_ms_, program_timeout_ms_));
+
+		for (;;)
+			Loop();
+	}
+
+private:
+	Connection& conn_;
+	const char* const name_;
+	std::optional<CoverFilter> max_signal_;
+	std::optional<CoverFilter> cover_filter_;
+	std::vector<std::unique_ptr<Proc>> procs_;
+	std::deque<rpc::ExecRequestRawT> requests_;
+	std::vector<std::string> leak_frames_;
+	uint32 slowdown_ = 0;
+	uint32 syscall_timeout_ms_ = 0;
+	uint32 program_timeout_ms_ = 0;
+
+	friend std::ostream& operator<<(std::ostream& ss, const Runner& runner)
+	{
+		ss << "procs:\n";
+		for (const auto& proc : runner.procs_)
+			ss << *proc;
+		ss << "\nqueued requests (" << runner.requests_.size() << "):\n";
+		for (const auto& req : runner.requests_)
+			ss << req;
+		return ss;
+	}
+
+	void Loop()
+	{
+		Select select;
+		select.Arm(conn_.FD());
+		for (auto& proc : procs_)
+			proc->Arm(select);
+		select.Wait(1000);
+		uint64 now = current_time_ms();
+
+		if (select.Ready(conn_.FD())) {
+			rpc::HostMessageRawT raw;
+			conn_.Recv(raw);
+			if (auto* msg = raw.msg.AsExecRequest())
+				Handle(*msg);
+			else if (auto* msg = raw.msg.AsSignalUpdate())
+				Handle(*msg);
+			else if (auto* msg = raw.msg.AsStartLeakChecks())
+				Handle(*msg);
+			else if (auto* msg = raw.msg.AsStateRequest())
+				Handle(*msg);
+			else
+				failmsg("unknown host message type", "type=%d", static_cast<int>(raw.msg.type));
+		}
+
+		for (auto& proc : procs_) {
+			proc->Ready(select, now, requests_.empty());
```

Comment:
If some child process has stalled, will such an approach make the whole prog execution on the VM wait for that proc to get repaired?

Reply:
All of these file descriptors are in non-blocking mode (set in Select::Prepare) and we use select to check which ones are ready, so shouldn't be the case.

Reply:
What exactly scenario do you mean?

Reply:
Actually any scenario when `proc->Ready()` blocks for some time.

E.g. if we are above the timeout and we want to restart the proc -- in that case we might have to wait for `Restart()`  to finish. `Restart()` can take a while, and in the mean time we will not be able to schedule anything to all other procs.

Reply:
It shouldn't block on anything. All file descriptors are in non-blocking mode.